File: ios/Plugin/MobileCronPlugin.swift
import Foundation
import Capacitor
import UIKit

@objc(MobileCronPlugin)
public class MobileCronPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MobileCronPlugin"
    public let jsName = "MobileCron"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "register", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unregister", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "update", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "list", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "triggerNow", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pauseAll", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resumeAll", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setMode", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testNativeEvaluate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testSetNextDueAt", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testInjectPendingEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testGetPendingCount", returnType: CAPPluginReturnPromise)
    ]

    private static let storageKey = "mobilecron:state"

    private var jobs: [String: [String: Any]] = [:]
    private var paused = false
    private(set) var mode = "balanced"
    var currentMode: String { mode }
    private var bgManager: BGTaskManager?

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    public override func load() {
        super.load()
        loadState()
        firePendingNativeEvents()
        let manager = BGTaskManager(plugin: self)
        manager.registerBGTasks()
        manager.scheduleRefresh()
        manager.scheduleProcessing(requiresExternalPower: true)
        self.bgManager = manager
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
    }

    @objc private func appDidBecomeActive() {
        firePendingNativeEvents()
    }

    func handleBackgroundWake(source: String) {
        firePendingNativeEvents()
        notifyListeners("statusChanged", data: buildStatus())
        notifyListeners("nativeWake", data: ["source": source, "paused": paused])
    }

    // ── Persistence ───────────────────────────────────────────────────────────

    private func loadState() {
        guard let raw = UserDefaults.standard.string(forKey: Self.storageKey),
              let data = raw.data(using: .utf8),
              let state = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return
        }
        paused = (state["paused"] as? Bool) ?? false
        mode = (state["mode"] as? String) ?? "balanced"
        jobs = [:]
        if let jobsArr = state["jobs"] as? [[String: Any]] {
            for job in jobsArr {
                if let id = job["id"] as? String, !id.isEmpty {
                    jobs[id] = job
                }
            }
        }
    }

    private func saveState() {
        var state: [String: Any] = [
            "version": 1,
            "paused": paused,
            "mode": mode,
            "jobs": Array(jobs.values)
        ]
        // Preserve any pendingNativeEvents written by NativeJobEvaluator
        if let raw = UserDefaults.standard.string(forKey: Self.storageKey),
           let data = raw.data(using: .utf8),
           let existing = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
           let pending = existing["pendingNativeEvents"] as? [[String: Any]],
           !pending.isEmpty {
            state["pendingNativeEvents"] = pending
        }
        guard let data = try? JSONSerialization.data(withJSONObject: state),
              let raw = String(data: data, encoding: .utf8) else { return }
        UserDefaults.standard.set(raw, forKey: Self.storageKey)
        // Force immediate disk flush so state survives simctl terminate / force-kill
        UserDefaults.standard.synchronize()
    }

    // ── Background wake ───────────────────────────────────────────────────────

    /// Read pendingNativeEvents from storage, emit each as jobDue, then clear them.
    private func firePendingNativeEvents() {
        guard let raw = UserDefaults.standard.string(forKey: Self.storageKey),
              let data = raw.data(using: .utf8),
              var state = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return
        }

        // Sync native-evaluated job fields (nextDueAt, lastFiredAt, consecutiveSkips) into memory.
        if let jobsArr = state["jobs"] as? [[String: Any]] {
            for nativeJob in jobsArr {
                if let id = nativeJob["id"] as? String, !id.isEmpty, jobs[id] != nil {
                    jobs[id] = nativeJob
                }
            }
        }

        guard let pending = state["pendingNativeEvents"] as? [[String: Any]], !pending.isEmpty else {
            return
        }

        for evt in pending {
            notifyListeners("jobDue", data: evt)
        }

        // Clear pendingNativeEvents and write back updated jobs.
        state.removeValue(forKey: "pendingNativeEvents")
        state["jobs"] = Array(jobs.values)
        if let newData = try? JSONSerialization.data(withJSONObject: state),
           let newRaw = String(data: newData, encoding: .utf8) {
            UserDefaults.standard.set(newRaw, forKey: Self.storageKey)
            UserDefaults.standard.synchronize()
        }
    }

    // ── Plugin methods ────────────────────────────────────────────────────────

    @objc func register(_ call: CAPPluginCall) {
        guard let name = call.getString("name")?.trimmingCharacters(in: .whitespacesAndNewlines), !name.isEmpty else {
            call.reject("Job name is required")
            return
        }

        let id = UUID().uuidString
        var record: [String: Any] = [
            "id": id,
            "name": name,
            "enabled": true,
            "consecutiveSkips": 0
        ]
        if let schedule = call.getObject("schedule") { record["schedule"] = schedule }
        if let activeHours = call.getObject("activeHours") { record["activeHours"] = activeHours }
        if call.options.keys.contains("requiresNetwork") { record["requiresNetwork"] = call.getBool("requiresNetwork") ?? false }
        if call.options.keys.contains("requiresCharging") { record["requiresCharging"] = call.getBool("requiresCharging") ?? false }
        if let priority = call.getString("priority") { record["priority"] = priority }
        if let data = call.getObject("data") { record["data"] = data }

        jobs[id] = record
        saveState()
        notifyListeners("statusChanged", data: buildStatus())
        call.resolve(["id": id])
    }

    @objc func unregister(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("id is required")
            return
        }
        jobs.removeValue(forKey: id)
        saveState()
        notifyListeners("statusChanged", data: buildStatus())
        call.resolve()
    }

    @objc func update(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("id is required")
            return
        }
        guard var existing = jobs[id] else {
            call.reject("Job not found")
            return
        }

        if let name = call.getString("name") { existing["name"] = name }
        if let schedule = call.getObject("schedule") { existing["schedule"] = schedule }
        if call.options.keys.contains("activeHours") { existing["activeHours"] = call.getObject("activeHours") }
        if call.options.keys.contains("requiresNetwork") { existing["requiresNetwork"] = call.getBool("requiresNetwork") ?? false }
        if call.options.keys.contains("requiresCharging") { existing["requiresCharging"] = call.getBool("requiresCharging") ?? false }
        if let priority = call.getString("priority") { existing["priority"] = priority }
        if call.options.keys.contains("data") { existing["data"] = call.getObject("data") }

        jobs[id] = existing
        saveState()
        notifyListeners("statusChanged", data: buildStatus())
        call.resolve()
    }

    @objc func list(_ call: CAPPluginCall) {
        call.resolve(["jobs": Array(jobs.values)])
    }

    @objc func triggerNow(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("id is required")
            return
        }
        guard let job = jobs[id] else {
            call.reject("Job not found")
            return
        }

        var payload: [String: Any] = [
            "id": id,
            "name": (job["name"] as? String) ?? "",
            "firedAt": Int(Date().timeIntervalSince1970 * 1000),
            "source": "manual"
        ]
        if let data = job["data"] {
            payload["data"] = data
        }
        notifyListeners("jobDue", data: payload)
        call.resolve()
    }

    @objc func pauseAll(_ call: CAPPluginCall) {
        paused = true
        saveState()
        notifyListeners("statusChanged", data: buildStatus())
        call.resolve()
    }

    @objc func resumeAll(_ call: CAPPluginCall) {
        paused = false
        saveState()
        notifyListeners("statusChanged", data: buildStatus())
        call.resolve()
    }

    @objc func setMode(_ call: CAPPluginCall) {
        guard let mode = call.getString("mode"), ["eco", "balanced", "aggressive"].contains(mode) else {
            call.reject("mode must be eco|balanced|aggressive")
            return
        }
        self.mode = mode
        if let bgManager {
            bgManager.scheduleRefresh()
            bgManager.scheduleProcessing(requiresExternalPower: mode != "aggressive")
        }
        saveState()
        notifyListeners("statusChanged", data: buildStatus())
        call.resolve()
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        call.resolve(buildStatus())
    }

    // ── E2E test hooks (not for production use) ───────────────────────────────

    /// Calls NativeJobEvaluator.evaluate() directly and fires pending events.
    @objc func testNativeEvaluate(_ call: CAPPluginCall) {
        let events = NativeJobEvaluator.evaluate(source: "test_trigger")
        firePendingNativeEvents()
        call.resolve(["firedCount": events.count])
    }

    /// Sets a job's nextDueAt in memory and saves to UserDefaults.
    /// Allows tests to mark a job as "due" without waiting for real time to pass.
    @objc func testSetNextDueAt(_ call: CAPPluginCall) {
        guard let id = call.getString("id"),
              let nextDueAtMs = call.getInt("nextDueAtMs"),
              jobs[id] != nil else {
            call.reject("id or nextDueAtMs missing, or job not found")
            return
        }
        jobs[id]?["nextDueAt"] = nextDueAtMs
        saveState()
        call.resolve()
    }

    /// Appends a pending native event to UserDefaults storage.
    /// Allows tests to simulate what NativeJobEvaluator writes during background execution.
    @objc func testInjectPendingEvent(_ call: CAPPluginCall) {
        guard let event = call.getObject("event") else {
            call.reject("event is required")
            return
        }
        guard let raw = UserDefaults.standard.string(forKey: Self.storageKey),
              let data = raw.data(using: .utf8),
              var state = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            call.reject("State not found in UserDefaults")
            return
        }
        var pending = (state["pendingNativeEvents"] as? [[String: Any]]) ?? []
        pending.append(event)
        state["pendingNativeEvents"] = pending
        guard let newData = try? JSONSerialization.data(withJSONObject: state),
              let newRaw = String(data: newData, encoding: .utf8) else {
            call.reject("Serialization failed")
            return
        }
        UserDefaults.standard.set(newRaw, forKey: Self.storageKey)
        UserDefaults.standard.synchronize()
        call.resolve()
    }

    /// Returns the count of pendingNativeEvents currently in UserDefaults storage.
    @objc func testGetPendingCount(_ call: CAPPluginCall) {
        guard let raw = UserDefaults.standard.string(forKey: Self.storageKey),
              let data = raw.data(using: .utf8),
              let state = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            call.resolve(["count": 0])
            return
        }
        let count = (state["pendingNativeEvents"] as? [[String: Any]])?.count ?? 0
        call.resolve(["count": count])
    }

    private func buildStatus() -> [String: Any] {
        let diagnostics = bgManager?.status ?? .init()
        return [
            "paused": paused,
            "mode": mode,
            "platform": "ios",
            "activeJobCount": jobs.count,
            "ios": [
                "bgRefreshRegistered": diagnostics.bgRefreshRegistered,
                "bgProcessingRegistered": diagnostics.bgProcessingRegistered,
                "bgContinuedAvailable": diagnostics.bgContinuedAvailable
            ]
        ]
    }
}

File: android/src/main/java/io/mobilecron/MobileCronPlugin.kt
package io.mobilecron

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.work.Constraints
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.ExistingWorkPolicy
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

@CapacitorPlugin(name = "MobileCron")
class MobileCronPlugin : Plugin() {
    private val jobs = ConcurrentHashMap<String, JSObject>()
    private var paused = false
    private var mode = "balanced"
    private var chargingReceiver: ChargingReceiver? = null

    companion object {
        private const val STORAGE_FILE = "CapacitorStorage"
        private const val STORAGE_KEY = "mobilecron:state"
    }

    override fun load() {
        super.load()
        CronBridge.plugin = this
        loadState()
        registerChargingReceiver()
        scheduleWorkManager()
    }

    override fun handleOnResume() {
        super.handleOnResume()
        // Fire any pendingNativeEvents written by NativeJobEvaluator while the app was backgrounded/killed.
        firePendingNativeEvents()
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        if (CronBridge.plugin === this) {
            CronBridge.plugin = null
        }
        chargingReceiver?.let {
            runCatching { context.unregisterReceiver(it) }
        }
        chargingReceiver = null
    }

    // ── Persistence ─────────────────────────────────────────────────────────

    private fun loadState() {
        val prefs = context.getSharedPreferences(STORAGE_FILE, Context.MODE_PRIVATE)
        val raw = prefs.getString(STORAGE_KEY, null) ?: return
        try {
            val json = JSONObject(raw)
            paused = json.optBoolean("paused", false)
            mode = json.optString("mode", "balanced")
            jobs.clear()
            val arr = json.optJSONArray("jobs") ?: return
            for (i in 0 until arr.length()) {
                val job = arr.optJSONObject(i) ?: continue
                val id = job.optString("id").takeIf { it.isNotEmpty() } ?: continue
                jobs[id] = JSObject.fromJSONObject(job)
            }
        } catch (_: Exception) { /* ignore corrupt state */ }
    }

    private fun saveState() {
        try {
            val state = JSONObject()
            state.put("version", 1)
            state.put("paused", paused)
            state.put("mode", mode)
            val arr = JSONArray()
            jobs.values.forEach { arr.put(JSONObject(it.toString())) }
            state.put("jobs", arr)

            // Preserve any pendingNativeEvents written by NativeJobEvaluator
            val prefs = context.getSharedPreferences(STORAGE_FILE, Context.MODE_PRIVATE)
            val existing = prefs.getString(STORAGE_KEY, null)
            if (existing != null) {
                val pending = runCatching { JSONObject(existing).optJSONArray("pendingNativeEvents") }.getOrNull()
                if (pending != null && pending.length() > 0) {
                    state.put("pendingNativeEvents", pending)
                }
            }

            prefs.edit().putString(STORAGE_KEY, state.toString()).apply()
        } catch (_: Exception) { /* ignore serialisation errors */ }
    }

    // ── Background wake ──────────────────────────────────────────────────────

    internal fun notifyFromBackground(source: String) {
        firePendingNativeEvents()
        val wakePayload = JSObject()
        wakePayload.put("source", source)
        wakePayload.put("paused", paused)
        notifyListeners("statusChanged", buildStatus())
        notifyListeners("nativeWake", wakePayload)
    }

    /** Read pendingNativeEvents from storage, emit each as jobDue, then clear them. */
    private fun firePendingNativeEvents() {
        val prefs = context.getSharedPreferences(STORAGE_FILE, Context.MODE_PRIVATE)
        val raw = prefs.getString(STORAGE_KEY, null) ?: return
        try {
            val json = JSONObject(raw)

            // Sync native-evaluated job fields (nextDueAt, lastFiredAt, consecutiveSkips) into memory.
            val jobsArr = json.optJSONArray("jobs")
            if (jobsArr != null) {
                for (i in 0 until jobsArr.length()) {
                    val nativeJob = jobsArr.optJSONObject(i) ?: continue
                    val id = nativeJob.optString("id").takeIf { it.isNotEmpty() } ?: continue
                    if (jobs.containsKey(id)) {
                        jobs[id] = JSObject.fromJSONObject(nativeJob)
                    }
                }
            }

            // Fire and clear pending native events.
            val pending = json.optJSONArray("pendingNativeEvents")
            if (pending == null || pending.length() == 0) return

            for (i in 0 until pending.length()) {
                val evt = pending.optJSONObject(i) ?: continue
                notifyListeners("jobDue", JSObject.fromJSONObject(evt))
            }

            // Clear pendingNativeEvents from storage.
            val cleared = JSONObject(raw)
            cleared.remove("pendingNativeEvents")
            val updatedArr = JSONArray()
            jobs.values.forEach { updatedArr.put(JSONObject(it.toString())) }
            cleared.put("jobs", updatedArr)
            prefs.edit().putString(STORAGE_KEY, cleared.toString()).apply()
        } catch (_: Exception) { /* ignore */ }
    }

    // ── WorkManager scheduling ───────────────────────────────────────────────

    private fun scheduleWorkManager() {
        val wm = WorkManager.getInstance(context)
        if (mode == "aggressive") {
            val request = OneTimeWorkRequestBuilder<CronChainWorker>()
                .setInitialDelay(5, TimeUnit.MINUTES)
                .build()
            wm.enqueueUniqueWork("mobilecron_chain", ExistingWorkPolicy.REPLACE, request)
            return
        }

        val constraintsBuilder = Constraints.Builder()
        if (mode == "eco") {
            constraintsBuilder.setRequiredNetworkType(NetworkType.UNMETERED)
            constraintsBuilder.setRequiresBatteryNotLow(true)
        } else {
            constraintsBuilder.setRequiredNetworkType(NetworkType.CONNECTED)
        }

        val request = PeriodicWorkRequestBuilder<CronWorker>(15, TimeUnit.MINUTES)
            .setConstraints(constraintsBuilder.build())
            .build()

        wm.enqueueUniquePeriodicWork(
            "mobilecron_periodic",
            ExistingPeriodicWorkPolicy.UPDATE,
            request
        )
    }

    private fun registerChargingReceiver() {
        if (chargingReceiver != null) return
        val receiver = ChargingReceiver()
        context.registerReceiver(receiver, IntentFilter(Intent.ACTION_POWER_CONNECTED))
        chargingReceiver = receiver
    }

    // ── Plugin methods ───────────────────────────────────────────────────────

    @PluginMethod
    fun register(call: PluginCall) {
        val name = call.getString("name")?.trim()
        if (name.isNullOrEmpty()) {
            call.reject("Job name is required")
            return
        }

        val id = UUID.randomUUID().toString()
        val record = JSObject()
        record.put("id", id)
        record.put("name", name)
        record.put("enabled", true)
        record.put("schedule", call.getObject("schedule") ?: JSObject())
        record.put("activeHours", call.getObject("activeHours"))
        record.put("requiresNetwork", call.getBoolean("requiresNetwork", false))
        record.put("requiresCharging", call.getBoolean("requiresCharging", false))
        record.put("priority", call.getString("priority", "normal"))
        call.getObject("data")?.let { record.put("data", it) }
        record.put("consecutiveSkips", 0)
        jobs[id] = record
        saveState()

        val result = JSObject()
        result.put("id", id)
        call.resolve(result)
        notifyListeners("statusChanged", buildStatus())
    }

    @PluginMethod
    fun unregister(call: PluginCall) {
        val id = call.getString("id")
        if (id == null) {
            call.reject("id is required")
            return
        }
        jobs.remove(id)
        saveState()
        call.resolve()
        notifyListeners("statusChanged", buildStatus())
    }

    @PluginMethod
    fun update(call: PluginCall) {
        val id = call.getString("id")
        if (id == null) {
            call.reject("id is required")
            return
        }
        val existing = jobs[id]
        if (existing == null) {
            call.reject("Job not found")
            return
        }

        call.getString("name")?.let { existing.put("name", it) }
        call.getObject("schedule")?.let { existing.put("schedule", it) }
        if (call.data.has("activeHours")) existing.put("activeHours", call.getObject("activeHours"))
        if (call.data.has("requiresNetwork")) existing.put("requiresNetwork", call.getBoolean("requiresNetwork", false))
        if (call.data.has("requiresCharging")) existing.put("requiresCharging", call.getBoolean("requiresCharging", false))
        call.getString("priority")?.let { existing.put("priority", it) }
        if (call.data.has("data")) existing.put("data", call.getObject("data"))

        jobs[id] = existing
        saveState()
        call.resolve()
        notifyListeners("statusChanged", buildStatus())
    }

    @PluginMethod
    fun list(call: PluginCall) {
        val arr = JSArray()
        jobs.values.forEach { arr.put(it) }
        val result = JSObject()
        result.put("jobs", arr)
        call.resolve(result)
    }

    @PluginMethod
    fun triggerNow(call: PluginCall) {
        val id = call.getString("id")
        if (id == null) {
            call.reject("id is required")
            return
        }
        val job = jobs[id]
        if (job == null) {
            call.reject("Job not found")
            return
        }

        val payload = JSObject()
        payload.put("id", id)
        payload.put("name", job.getString("name"))
        payload.put("firedAt", System.currentTimeMillis())
        payload.put("source", "manual")
        if (job.has("data")) payload.put("data", job.getJSONObject("data"))
        notifyListeners("jobDue", payload)
        call.resolve()
    }

    @PluginMethod
    fun pauseAll(call: PluginCall) {
        paused = true
        saveState()
        call.resolve()
        notifyListeners("statusChanged", buildStatus())
    }

    @PluginMethod
    fun resumeAll(call: PluginCall) {
        paused = false
        saveState()
        call.resolve()
        notifyListeners("statusChanged", buildStatus())
    }

    @PluginMethod
    fun setMode(call: PluginCall) {
        val next = call.getString("mode")
        if (next !in listOf("eco", "balanced", "aggressive")) {
            call.reject("mode must be eco|balanced|aggressive")
            return
        }
        mode = next!!
        scheduleWorkManager()
        saveState()
        call.resolve()
        notifyListeners("statusChanged", buildStatus())
    }

    @PluginMethod
    fun getStatus(call: PluginCall) {
        call.resolve(buildStatus())
    }

    private fun buildStatus(): JSObject {
        val status = JSObject()
        status.put("paused", paused)
        status.put("mode", mode)
        status.put("platform", "android")
        status.put("activeJobCount", jobs.size)
        status.put("android", JSObject().apply {
            put("workManagerActive", true)
            put("chargingReceiverActive", chargingReceiver != null)
        })
        return status
    }
}
